### PR TITLE
[cli] Add automatic gas estimation, and move all CLI client commands to use it

### PR DESCRIPTION
**Pull request by stefan-mysten**
## Description 

This PR adds automatic gas estimation support for all CLI client commands by default. Providing a specific gas budget is still possible through the `--gas-budget` option.

Note this also fixes a problem with the way CLI PTB tests are executed, which led to issues (cannot find test files).

## Test Plan 

Added new tests + existing tests.
`cargo nextest run --test cli_tests`

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
The CLI client commands do not require to provide a `--gas-budget` value anymore, as it is automatically inferred through a dry run. It is still possible to provide one, if the user wants.

### CODE REVIEW

**Review by amnn — COMMENTED** [review]
I like the change overall! Main thing I'd love to change before we land this is how the transactions are created.

Today, you are duplicating the logic involved in creating the `TransactionData`, once for the gas estimation and another time to actually run it. This is not great from a code repetition perspective, but it's also not good from an efficiency or consistency perspective. Some of these transaction building commands do a fair bit of work, talking to RPC to resolve ObjectIDs into ObjectRefs etc, you don't want to redo that work because it's not cheap, and in general you may get a different answer from run to run, and while that's usually going to mean the transaction execution will fail, we should still return a consistent response in that case (e.g. the transaction you dry ran is the same as the one that you executed).

The fix is to split the logic for generating transactions into three phases:

- Build the `TransactionKind` -- this is usually where most of the effort and consistency concerns go, and we can always do this once because the gas budget is not relevant for this part.
- Build the `TransactionData` for dry-run, from a clone of the `TransactionKind`.
- Build the `TransactionData` for execution, using the gas budget you got from dry-run, and execute it.

File: crates/sui/src/client_commands.rs [hunk 1]
```diff
@@ -2432,3 +2598,26 @@ fn pretty_print_balance(
         .with(TableStyle::blank())
         .to_string()]);
 }
+
+pub(crate) async fn dry_run(
```
[the remark this answers is not in the document]
Comment (stefan-mysten) [reply crates/sui/src/client_commands.rs:2602]: 🤦 of course.

**Comment by stefan-mysten** [comment]
> I like the change overall! Main thing I'd love to change before we land this is how the transactions are created.
> 
> Today, you are duplicating the logic involved in creating the `TransactionData`, once for the gas estimation and another time to actually run it. This is not great from a code repetition perspective, but it's also not good from an efficiency or consistency perspective. Some of these transaction building commands do a fair bit of work, talking to RPC to resolve ObjectIDs into ObjectRefs etc, you don't want to redo that work because it's not cheap, and in general you may get a different answer from run to run, and while that's usually going to mean the transaction execution will fail, we should still return a consistent response in that case (e.g. the transaction you dry ran is the same as the one that you executed).
> 
> The fix is to split the logic for generating transactions into three phases:
> 
>     * Build the `TransactionKind` -- this is usually where most of the effort and consistency concerns go, and we can always do this once because the gas budget is not relevant for this part.
> 
>     * Build the `TransactionData` for dry-run, from a clone of the `TransactionKind`.
> 
>     * Build the `TransactionData` for execution, using the gas budget you got from dry-run, and execute it.

Agreed, I was quite unhappy with the way I did it. Will refactor!

**Comment by stefan-mysten** [comment]
Superseded by #17322

**Pull request closed without merging**